In [ ]:
import pandas as pd

### Reading in the data

In [ ]:
data_2014 = pd.read_csv('Analytics_mindset_case_studies_PCard_FY2014.csv')

In [ ]:
data_2015 = pd.read_csv('Analytics_mindset_case_studies_PCard_FY2015.csv')

/Users/davidbianchin/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (7,8) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


### Looking at the data

In [ ]:
data_2014.head()

Agency Number                Agency Name Cardholder Last Name  \
0           1000  OKLAHOMA STATE UNIVERSITY                Mason   
1           1000  OKLAHOMA STATE UNIVERSITY                Mason   
2           1000  OKLAHOMA STATE UNIVERSITY               Massey   
3           1000  OKLAHOMA STATE UNIVERSITY               Massey   
4           1000  OKLAHOMA STATE UNIVERSITY        Mauro-Herrera   

  Cardholder First Initial                   Description  Amount  \
0                        C              GENERAL PURCHASE  890.00   
1                        C                  ROOM CHARGES  368.96   
2                        J              GENERAL PURCHASE  165.82   
3                        T              GENERAL PURCHASE   96.39   
4                        M  HAMMERMILL COPY PLUS COPY EA  125.96   

           Vendor Transaction Date     Posted Date  \
0           NACAS   7/30/2013 0:00  7/31/2013 0:00   
1  SHERATON HOTEL   7/30/2013 0:00  7/31/2013 0:00   
2  SEARS.COM 9300   7/29/2013 0:00  7/31/2013 0:00   
3  WAL-MART #0137   7/30/2013 0:00  7/31/2013 0:00   
4  STAPLES DIRECT   7/30/2013 0:00  7/31/2013 0:00   

                        Merchant Category Code (MCC)  
0        CHARITABLE AND SOCIAL SERVICE ORGANIZATIONS  
1                                           SHERATON  
2  DIRCT MARKETING/DIRCT MARKETERS--NOT ELSEWHERE...  
3                    GROCERY STORES,AND SUPERMARKETS  
4  STATIONERY, OFFICE SUPPLIES, PRINTING AND WRIT...

In [ ]:
data_2015.head()

Agency Number                Agency Name Cardholder Last Name  \
0           1000  OKLAHOMA STATE UNIVERSITY                Clark   
1           1000  OKLAHOMA STATE UNIVERSITY                Clark   
2           1000  OKLAHOMA STATE UNIVERSITY                Clark   
3           1000  OKLAHOMA STATE UNIVERSITY                Clark   
4           1000  OKLAHOMA STATE UNIVERSITY                Clark   

  Cardholder First Initial                                        Description  \
0                        K                                   GENERAL PURCHASE   
1                        M                                   GENERAL PURCHASE   
2                        M                              NEWS ADV./SUBSC. EACH   
3                        M                                   GENERAL PURCHASE   
4                        P  BH 3-INX7-IN SMALL BRASS B PCS|BLUE HAWK SHOE ...   

       Amount                  Vendor Transaction Date     Posted Date  \
0     $88.88   Discount School Supply   7/23/2014 0:00  7/24/2014 0:00   
1    $425.00   BROONER'S FLOOR CENTER   7/23/2014 0:00  7/24/2014 0:00   
2    $141.50           820 NEWS PRESS   7/23/2014 0:00  7/24/2014 0:00   
3  $2,679.99           WWW.NEWEGG.COM   7/23/2014 0:00  7/24/2014 0:00   
4     $22.44             LOWES #00241   7/22/2014 0:00  7/24/2014 0:00   

                 Merchant Category Code (MCC)  
0  STATIONERY,OFFICE AND SCHOOL SUPPLY STORES  
1        FLOOR COVERING,RUG AND CARPET STORES  
2           CONTINUITY/SUBSCRIPTION MERCHANTS  
3                          ELECTRONICS STORES  
4                HOME SUPPLY WAREHOUSE STORES

In [ ]:
data_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442458 entries, 0 to 442457
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Agency Number                 442458 non-null  int64  
 1   Agency Name                   442458 non-null  object 
 2   Cardholder Last Name          442458 non-null  object 
 3   Cardholder First Initial      442458 non-null  object 
 4   Description                   442458 non-null  object 
 5   Amount                        442458 non-null  float64
 6   Vendor                        442458 non-null  object 
 7   Transaction Date              442458 non-null  object 
 8   Posted Date                   442458 non-null  object 
 9   Merchant Category Code (MCC)  442458 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 33.8+ MB


In [ ]:
data_2015.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 437579 entries, 0 to 437578
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   Agency Number                 437579 non-null  int64 
 1   Agency Name                   437579 non-null  object
 2   Cardholder Last Name          437579 non-null  object
 3   Cardholder First Initial      437579 non-null  object
 4   Description                   437579 non-null  object
 5   Amount                        437579 non-null  object
 6   Vendor                        437579 non-null  object
 7   Transaction Date              214698 non-null  object
 8   Posted Date                   214698 non-null  object
 9   Merchant Category Code (MCC)  437579 non-null  object
dtypes: int64(1), object(9)
memory usage: 33.4+ MB


#### Notes:
- The amounts in 2015 have the dollar symbol while 2014 is classified as float, also they have commas and "()"
- Lost of nulls in Transaction Date and posted Date in 2015
- Column names have spaces

### Transforming Data in original Files

In [169]:
# Converting the Amount column to float
data_2015['Amount'] = data_2015['Amount'].str.replace('[\$,)]', '', regex=True)

# Changing (5.15) to -5.15
data_2015['Amount'] = data_2015['Amount'].str.replace('(', '-', regex = True)

data_2015['Amount'] = data_2015['Amount'].astype(float)

In [170]:
data_2015.dtypes

Agency Number                     int64
Agency Name                      object
Cardholder Last Name             object
Cardholder First Initial         object
Description                      object
Amount                          float64
Vendor                           object
Transaction Date                 object
Posted Date                      object
Merchant Category Code (MCC)     object
dtype: object

In [171]:
# Removing WhiteSpaces from the column names before merging -> as per instructions

# remove spaces from column names of data_2014
data_2014.rename(columns=lambda x: x.replace(' ', ''), inplace=True)

# remove spaces from column names of data_2015
data_2015.rename(columns=lambda x: x.replace(' ', ''), inplace=True)

In [172]:
dataset = pd.concat([data_2014, data_2015]).reset_index(drop=True)

In [173]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880037 entries, 0 to 880036
Data columns (total 10 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   AgencyNumber               880037 non-null  int64  
 1   AgencyName                 880037 non-null  object 
 2   CardholderLastName         880037 non-null  object 
 3   CardholderFirstInitial     880037 non-null  object 
 4   Description                880037 non-null  object 
 5   Amount                     880037 non-null  float64
 6   Vendor                     880037 non-null  object 
 7   TransactionDate            657156 non-null  object 
 8   PostedDate                 657156 non-null  object 
 9   MerchantCategoryCode(MCC)  880037 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 67.1+ MB


In [174]:
dataset['TransactionDate'] = pd.to_datetime(dataset['TransactionDate'], format='%m/%d/%Y %H:%M')
dataset['PostedDate'] = pd.to_datetime(dataset['PostedDate'], format='%m/%d/%Y %H:%M')

# All times are 0:00
# Not sure if I should set NaT's to ''

In [175]:
# Removing trailing and leading spaces from data
dataset = dataset.apply(lambda x: x.str.strip() if x.dtype == "object" else x)


In [176]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 880037 entries, 0 to 880036
Data columns (total 10 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   AgencyNumber               880037 non-null  int64         
 1   AgencyName                 880037 non-null  object        
 2   CardholderLastName         880037 non-null  object        
 3   CardholderFirstInitial     880037 non-null  object        
 4   Description                880037 non-null  object        
 5   Amount                     880037 non-null  float64       
 6   Vendor                     880037 non-null  object        
 7   TransactionDate            657156 non-null  datetime64[ns]
 8   PostedDate                 657156 non-null  datetime64[ns]
 9   MerchantCategoryCode(MCC)  880037 non-null  object        
dtypes: datetime64[ns](2), float64(1), int64(1), object(6)
memory usage: 67.1+ MB


In [177]:
dataset.head()

AgencyNumber                 AgencyName CardholderLastName  \
0          1000  OKLAHOMA STATE UNIVERSITY              Mason   
1          1000  OKLAHOMA STATE UNIVERSITY              Mason   
2          1000  OKLAHOMA STATE UNIVERSITY             Massey   
3          1000  OKLAHOMA STATE UNIVERSITY             Massey   
4          1000  OKLAHOMA STATE UNIVERSITY      Mauro-Herrera   

  CardholderFirstInitial                   Description  Amount  \
0                      C              GENERAL PURCHASE  890.00   
1                      C                  ROOM CHARGES  368.96   
2                      J              GENERAL PURCHASE  165.82   
3                      T              GENERAL PURCHASE   96.39   
4                      M  HAMMERMILL COPY PLUS COPY EA  125.96   

           Vendor TransactionDate PostedDate  \
0           NACAS      2013-07-30 2013-07-31   
1  SHERATON HOTEL      2013-07-30 2013-07-31   
2  SEARS.COM 9300      2013-07-29 2013-07-31   
3  WAL-MART #0137      2013-07-30 2013-07-31   
4  STAPLES DIRECT      2013-07-30 2013-07-31   

                           MerchantCategoryCode(MCC)  
0        CHARITABLE AND SOCIAL SERVICE ORGANIZATIONS  
1                                           SHERATON  
2  DIRCT MARKETING/DIRCT MARKETERS--NOT ELSEWHERE...  
3                    GROCERY STORES,AND SUPERMARKETS  
4  STATIONERY, OFFICE SUPPLIES, PRINTING AND WRIT...

In [182]:
# dataset.to_csv('task3_final_dataset.csv', index = False)

### ANSWERS

##### There are 880037 rows in the final file
##### The total sum of the amount column is 379,101,777.6800001

In [178]:
dataset['Amount'].sum()

# Double checked in excel and the sum is correct

379101777.6800001

In [179]:
dataset.shape

(880037, 10)

In [ ]:
dataset.shape

In [ ]:
#dataset for OSU 204
o_data=dataset